runtime/lua/vim/diagnostic.lua

```diff
@@ -1603,6 +1616,271 @@ M.handlers.virtual_text = {
   end,
 }
 
+local virtual_lines_augroup =
+  api.nvim_create_augroup('vim_lsp_diagnostic_virt_lines', { clear = true })
+
+---Some characters (like tabs) take up more than one cell. Additionally, inline
+---virtual text can make the distance between 2 columns larger.
+---A diagnostic aligned under such characters needs to account for that and that
+---many spaces to its left.
+---@param bufnr integer
+---@param lnum integer
+---@param start_col integer
+---@param end_col integer
+---@return integer
+local function distance_between_cols(bufnr, lnum, start_col, end_col)
+  return api.nvim_buf_call(bufnr, function()
+    local s = vim.fn.virtcol({ lnum + 1, start_col })
+    local e = vim.fn.virtcol({ lnum + 1, end_col + 1 })
+    return e - 1 - s
+  end)
+end
+
+---@param namespace integer
+---@param bufnr integer
+---@param diagnostics vim.Diagnostic[]
+---@param opts vim.diagnostic.Opts.VirtualLines
+local function render_virtual_lines(namespace, bufnr, diagnostics, opts)
+  if not api.nvim_buf_is_loaded(bufnr) then
+    return
+  end
+
+  vim.validate('namespace', namespace, 'number')
+  vim.validate('bufnr', bufnr, 'number')
+  vim.validate('diagnostics', diagnostics, vim.islist, 'a list of diagnostics')
+  vim.validate('opts', opts, 'table', true)
```

Comment:
We shouldn't need to validate in a local (internal) function. We have full control over the inputs to these functions so should assume all parameters are valid. Validation should only occur in public/user-facing functions.

We can also assume that the buffer is loaded at this point so can skip the `buf_is_loaded` check.